src/server/services/records.js

```diff
@@ -120,9 +125,43 @@ async function recordsForReportingPeriod (tenantId, periodId) {
   return groupedRecords.flat()
 }
 
+/**
+ * Get the most recent, validated record for each unique project, as of the
+ * specified reporting period.
+*/
+async function mostRecentProjectRecords (tenantId, periodId) {
+  log('mostRecentProjectRecords()')
+  requiredArgument(tenantId, 'must specify tenantId in mostRecentProjectRecords')
+  requiredArgument(periodId, 'must specify periodId in mostRecentProjectRecords')
+
+  const reportingPeriods = await getPreviousReportingPeriods(tenantId, periodId)
+
+  const allRecords = await Promise.all(
+    reportingPeriods.map(reportingPeriod =>
+      recordsForReportingPeriod(tenantId, reportingPeriod.id)
+    )
+  )
+
+  const latestProjectRecords = allRecords
+    .flat()
+    // exclude non-project records
+    .filter(record => Object.values(EC_SHEET_TYPES).includes(record.type))
+    // collect the latest record for each project ID
+    .reduce(
+      (accumulator, record) => ({
+        ...accumulator,
```

Comment:
I'm not super familiar with the js spread operator, but is it right to assume that each time this reduce callback is called it has to create a whole new object copied from the previous object, add/overwrite an entry, and then return it? Would that make this an O(N^2) iteration, instead of just O(N) if you did this instead?
```
.reduce(
  (accumulator, record) => {
    accumulator[record.content.Project_Identification_Number__c] = record
    return accumulator
})
```

Reply:
Yes, this is an accurate take!

However, in my experience, the cost of object manipulation in time and memory is negligible compared to performance costs like the I/O issue you identified above.

I like to default to this spread pattern just because it avoids any possibility of accidental non-local mutation.

To be fair, I don't think either accidental mutation or performance overhead are likely to be meaningful here.  😅